File: Week_06/403/frog_jump2.py
from typing import List


class Solution:
    def canCross(self, stones: List[int]) -> bool:
        if stones[1] != 1:
            return False
        n = len(stones)

        dp = [[False] * n for _ in range(n)]
        dp[0][0] = True
        for i in range(1, n):
            for j in range(0, i):
                k = stones[i] - stones[j]
                if k < 1 or k > j + 1:
                    continue
                dp[i][k] = dp[j][k - 1] or dp[j][k] or dp[j][k + 1]
                if i == n - 1 and dp[i][k]:
                    return True
        return False

File: Week_06/403/frog_jump2.go
package main

func canCross2(stones []int) bool {
	n := len(stones)
	dp := make([][]bool, n)
	for i := 0; i < n; i++ {
		dp[i] = make([]bool, n)
	}
	dp[0][0] = true
	for i := 1; i < n; i++ {
		for j := 0; j < i; j++ {
			k := stones[i]-stones[j]
			if k < 1 || k > j+1 || !dp[j][k]{
				continue
			}
			dp[i][k] = dp[j][k] || dp[j][k-1] || dp[j][k+1]
			if i == n-1  && dp[i][k] {
				return true
			}
		}
	}
	return false
}
